arch/x86/kvm/pkvm/mmu.c

```
@@ -612,3 +624,115 @@ int pkvm_hyp_donate_host_mmio_locked(unsigned long phys, unsigned long size)
 
 	return pkvm_pgtable_map(&host_mmu, phys, phys, size, prot);
 }
+
+/**
+ * pkvm_host_share_hyp() - Shares pages from host to hypervisor.
+ * @phys:	Physical address of the memory region to share.
+ * @size:	Size of the memory region to share.
+ *
+ * The sharing transfers the memory pages in range [@phys, @phys + @size) which
+ * are exclusively owned by the host to shared-owned by the host and the pKVM
+ * hypervisor. The host_share_hyp_count counter of those memory pages are
+ * incremented. If a memory page is already shared-owned, page state will not be
+ * changed but only increment the counter. @phys and @size are not required to be
+ * PAGE_SIZE aligned (@size is required to be non-zeroed value) to support
+ * sharing page-unaligned memory pages.
+ *
+ * Returns: 0 on success, or a negative error code on failure.
+ */
+int pkvm_host_share_hyp(unsigned long phys, unsigned long size)
+{
+	unsigned long end = PAGE_ALIGN(phys + size);
+	enum pkvm_page_state pgstates[] = {
+		PKVM_PAGE_OWNED,
+		PKVM_PAGE_SHARED_OWNED,
+	};
+	struct pkvm_page *page;
+	int ret;
+
+	if (size == 0)
+		return -EINVAL;
+
+	pkvm_host_mmu_lock();
+
+	ret = check_host_mem_pgstates(phys, size, pgstates, ARRAY_SIZE(pgstates));
+	if (ret)
+		goto unlock;
+
+	/*
+	 * TODO: Prevent sharing a memory page which is
+	 * shared-owned with a guest VM.
+	 */
+
+	for (phys = PAGE_ALIGN_DOWN(phys); phys < end; phys += PAGE_SIZE) {
+		page = pkvm_phys_to_page(phys);
+
+		if (page->host_state == PKVM_PAGE_OWNED)
+			page->host_state = PKVM_PAGE_SHARED_OWNED;
+
+		/*
+		 * The page state should be shared-owned otherwise
+		 * it must be a code bug.
+		 */
+		BUG_ON(page->host_state != PKVM_PAGE_SHARED_OWNED);
+
+		BUG_ON(page->host_share_hyp_count == U16_MAX);
```

Why BUG_ON? We can just return an error, like pKVM-ARM does?

The error value could be e.g. -ENOSPC (although pKVM-ARM returns -EBUSY in this case, but -ENOSPC seems more suitable?)

BTW also, like in pKVM-ARM, we could first check all pages for overflow, and only after that update the page states and counters, in a separate loop (to avoid failing in the middle of updating).

And while at it: pKVM-ARM also checks if host_share_guest_count is non-zero if current state is PKVM_PAGE_SHARED_OWNED, which is actually critical, since it prevents sharing a page with a guest if it is shared with the hypervisor. In our case here, while we have no guests yet, host_state == PKVM_PAGE_SHARED_OWNED && host_share_hyp_count == 0 may only occur due to a bug in pKVM, but once we have guests, it may also occur if the host  tries to share a page with hypervisor while it is already shared with a guest. So we could already add returning -EPERM in this case?

---

> ```
> +			/*
> +			 * TODO: for now host_share_hyp_count may only
> +			 * be 0 due to a pKVM bug. Once guest VM support
> +			 * is added, it may also happen if the host tries
> +			 * to share a page that is shared with a guest,
> +			 * so this BUG_ON should be changed to returning
> +			 * -EPERM then.
> +			 */
> +			BUG_ON(!page->host_share_hyp_count);
> ```

Or maybe not leave this TODO but do it the final way right away:
```
		/*
		 * Allow sharing an already shared page as long as it is shared
		 * with the hypervisor, not with a guest.
		 */
		if (page->host_share_hyp_count)
			continue;
```

---

/*
		 * Allow sharing an already shared page as long as it is shared
		 * with the hypervisor, not with a guest.
		 */
		if (page->host_share_hyp_count)
			continue;

This is smart, so we don't need to put a TODO. Will apply your change together with this one.

Noticed, we will unconditionally `page->host_state = PKVM_PAGE_SHARED_OWNED;` even it is already shared owned. I think it is fine as otherwise we will add additional lines to check, which seems doesn't save cycles.

Will also remove check_host_mem_pgstates().